### PR TITLE
Add recipe for HuBERT model pre-training

### PR DESCRIPTION
**Pull request by nateanl**
Replace #2129

### CODE REVIEW

File: examples/hubert/lightning.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
+from typing import Tuple
+
+import torch
+import torchaudio
+from dataset import BucketizeSampler, DistributedBatchSampler, HuBERTDataSet, CollateFnHubert
```
Comment (kobenaxie) [inline examples/hubert/lightning.py:5]: BucketizeSampler not defined.

Comment (nateanl) [reply examples/hubert/lightning.py:5]: Hi @kobenaxie, thanks for pointing it out. I recently updated the `BucketizeSampler` to `BucketizeBatchSampler` in https://github.com/pytorch/audio/blob/main/examples/hubert/dataset/hubert_dataset.py#L11. I will update this PR soon.

File: examples/hubert/loss/hubert_loss.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    masked_weight: float = 1.0,
+    nomask_weight: float = 0.0,
+    feature_weight: float = 10.0,
+    reduction: str = "sum",
+) -> Tensor:
+    """Compute the cross-entropy loss on HuBERT masked and non-masked logits.
+    Args:
+        logit_m (Tensor or None): The masked logit Tensor of dimension `[masked_frames, final_dim]`.
+        logit_u (Tensor or None): The non-masked logit Tensor of dimension `[nonmasked_frames, final_dim]`.
+        feature_pen (Tensor): The feature mean value for additional penalty loss.
```
Comment (hwangjeff) [inline examples/hubert/loss/hubert_loss.py:21]: i might prefer just writing out "feature_penalty" for clarity's sake

File: examples/hubert/loss/hubert_loss.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    logit_u: Optional[Tensor],
+    feature_pen: Tensor,
+    masked_weight: float = 1.0,
+    nomask_weight: float = 0.0,
+    feature_weight: float = 10.0,
+    reduction: str = "sum",
+) -> Tensor:
+    """Compute the cross-entropy loss on HuBERT masked and non-masked logits.
+    Args:
+        logit_m (Tensor or None): The masked logit Tensor of dimension `[masked_frames, final_dim]`.
```
Comment (hwangjeff) [inline examples/hubert/loss/hubert_loss.py:19]: should we standardize using tuples to represent shapes? e.g. `(masked_frames, final_dim)`

File: examples/hubert/loss/hubert_loss.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        logit_u (Tensor or None): The non-masked logit Tensor of dimension `[nonmasked_frames, final_dim]`.
+        feature_pen (Tensor): The feature mean value for additional penalty loss.
+        masked_weight (float, optional): The weight for masked cross-entropy loss (Default: ``1.0``).
+        nomask_weight (float, optional): The weight for non-masked cross-entropy loss (Default: ``0.0``).
```
Comment (hwangjeff) [inline examples/hubert/loss/hubert_loss.py:23]: `unmasked_weight`?

File: examples/hubert/train.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+from lightning import HuBERTPreTrainModule
+from pytorch_lightning import Trainer
+from pytorch_lightning.callbacks import ModelCheckpoint
+
+
+logger = logging.getLogger(__name__)
+
+
+class _Formatter(ArgumentDefaultsHelpFormatter, RawDescriptionHelpFormatter):
+    # https://stackoverflow.com/a/18462760
```
Comment (hwangjeff) [inline examples/hubert/train.py:21]: maybe a little more here so that readers can quickly get a sense of the context before clicking through for more details

**Review by hwangjeff — APPROVED** [review]

File: examples/hubert/loss/hubert_loss.py (hunk 3 above)
Comment (nateanl) [reply examples/hubert/loss/hubert_loss.py:19]: sure, will change it to `(masked_frames, final_dim)`
